### PR TITLE
Reintroduce TTS WS

### PR DESCRIPTION
**Pull request by dvonthenen**
## Proposed changes

Reintroduce TTS WS

## Types of changes

What types of changes does your code introduce to the community Python SDK?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update or tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I have lint'ed all of my code using repo standards
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Summary by CodeRabbit

- **New Features**
	- Introduced WebSocket capabilities for real-time interactions within the Deepgram client.
	- Added classes for handling WebSocket communications, including `SpeakWebSocketClient` and `AsyncSpeakWebSocketClient`.
	- Implemented various response classes to manage messages received from the WebSocket API, enhancing usability.
	- Created new example applications demonstrating asynchronous and interactive text-to-speech functionalities using WebSockets.

- **Improvements**
	- Improved organization of existing import statements for clarity and maintainability.
	- Enhanced functionality of options related to WebSocket interactions, allowing for better configuration.
	- Added necessary dependencies for example applications to function correctly.

### CODE REVIEW

File: examples/text-to-speech/websocket/interactive/main.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        # Create a websocket connection to Deepgram
+        dg_connection = deepgram.speak.websocket.v("1")
+        # print(dg_connection)
+
+        def on_open(self, open, **kwargs):
+            print(f"\n\n{open}\n\n")
+            thread = threading.Thread(target=send_tts_text, args=(self,))
+            thread.start()
+            thread.join()
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/interactive/main.py]: This has been fixed.

File: examples/text-to-speech/websocket/interactive/main.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        dg_connection.finish()
+
+        print("Finished")
+
+    except ConnectionClosedError as e:
+        print(f"WebSocket connection closed unexpectedly: {e}")
+    except ValueError as e:
+        print(f"Invalid value encountered: {e}")
+    except Exception as e:
+        print(f"An unexpected error occurred: {e}")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/interactive/main.py]: This is only an example.

File: examples/text-to-speech/websocket/interactive/main.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            print(f"\n\n{open}\n\n")
+            thread = threading.Thread(target=send_tts_text, args=(self,))
+            thread.start()
+            thread.join()
+
+        def on_binary_data(self, data, **kwargs):
+            print("Received binary data:")
+            with open(AUDIO_FILE, "ab") as f:
+                f.write(data)
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/interactive/main.py]: This is only an example.

File: examples/text-to-speech/websocket/async_interactive/main.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        # Create a websocket connection to Deepgram
+        dg_connection = deepgram.speak.asyncwebsocket.v("1")
+
+        async def on_open(client, open_response, **kwargs):
+            print(f"\n\nOpen: {open_response}\n\n")
+            await send_tts_text(client)
+
+        async def on_binary_data(client, data, **kwargs):
+            print("Received binary data")
+            await write_binary_to_mp3(data)
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/async_interactive/main.py]: This is only an example.

File: examples/text-to-speech/websocket/async_interactive/main.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        print("Finished")
+
+    except ConnectionClosedError as e:
+        print(f"WebSocket connection closed unexpectedly: {e}")
+    except asyncio.CancelledError as e:
+        print(f"Asyncio task was cancelled: {e}")
+    except OSError as e:
+        print(f"File operation failed: {e}")
+    except Exception as e:
+        print(f"An unexpected error occurred: {e}")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/async_interactive/main.py]: This is only an example.

File: deepgram/clients/speak/v1/websocket/client.py [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+    _listen_thread: Union[threading.Thread, None]
+
+    _kwargs: Optional[Dict] = None
+    _addons: Optional[Dict] = None
+    _options: Optional[Dict] = None
+    _headers: Optional[Dict] = None
+
+    def __init__(self, config: DeepgramClientOptions):
+        if config is None:
+            raise DeepgramError("Config are required")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/client.py:61]: Will address this in a different PR.

File: deepgram/clients/speak/v1/websocket/response.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    def __getitem__(self, key):
+        _dict = self.to_dict()
+        return _dict[key]
+
+    def __setitem__(self, key, val):
+        self.__dict__[key] = val
+
+    def __str__(self) -> str:
+        return self.to_json(indent=4)
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/response.py:63]: Will address this in a different PR.

File: deepgram/clients/speak/v1/websocket/response.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/response.py:101]: Will address this in a different PR.

File: deepgram/clients/speak/v1/websocket/response.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    def __setitem__(self, key, val):
+        self.__dict__[key] = val
+
+    # this is a hack to make the response look like a dict because of the io.BytesIO object
+    # otherwise it will throw an exception on printing
+    def __str__(self) -> str:
+        my_dict = self.to_dict()
+        return my_dict.__str__()
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/response.py:44]: Will address this in a different PR.

File: deepgram/clients/speak/v1/websocket/async_client.py [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+    _listen_thread: Union[asyncio.Task, None]
+
+    _kwargs: Optional[Dict] = None
+    _addons: Optional[Dict] = None
+    _options: Optional[Dict] = None
+    _headers: Optional[Dict] = None
+
+    def __init__(self, config: DeepgramClientOptions):
+        if config is None:
+            raise DeepgramError("Config are required")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/async_client.py:58]: Will address this in a different PR.

File: deepgram/clients/speak/v1/websocket/response.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    def __getitem__(self, key):
+        _dict = self.to_dict()
+        return _dict[key]
+
+    def __setitem__(self, key, val):
+        self.__dict__[key] = val
+
+    def __str__(self) -> str:
+        return self.to_json(indent=4)
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/response.py:82]: Will address this in a different PR.

File: examples/text-to-speech/websocket/async_complete/main.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+        async def on_warning(self, warning, **kwargs):
+            print(f"\n\n{warning}\n\n")
+
+        async def on_error(self, error, **kwargs):
+            print(f"\n\n{error}\n\n")
+
+        async def on_unhandled(self, unhandled, **kwargs):
+            print(f"\n\n{unhandled}\n\n")
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/async_complete/main.py:58]: This is just an example.

File: examples/text-to-speech/websocket/async_complete/main.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+            model="aura-asteria-en",
+            encoding="linear16",
+            container="none",
+            sample_rate=48000,
+        )
+
+        print("\n\nPress Enter to stop...\n\n")
+        if await dg_connection.start(options) is False:
+            print("Failed to start connection")
+            return
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/async_complete/main.py:79]: This is just an example.

File: examples/text-to-speech/websocket/async_complete/main.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+TTS_TEXT = "Hello, this is a text to speech example using Deepgram."
+
+
+async def main():
+    try:
+        # example of setting up a client config. logging values: WARNING, VERBOSE, DEBUG, SPAM
+        # config: DeepgramClientOptions = DeepgramClientOptions(verbose=verboselogs.DEBUG)
+        # deepgram: DeepgramClient = DeepgramClient("", config)
+        # otherwise, use default config
+        deepgram: DeepgramClient = DeepgramClient()
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/async_complete/main.py:27]: This is just an example.

File: examples/text-to-speech/websocket/async_complete/main.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        # Indicate that we've finished
+        await asyncio.sleep(6)
+        await dg_connection.finish()
+
+        print("Finished")
+
+    except ValueError as e:
+        print(f"Invalid value encountered: {e}")
+    except Exception as e:
+        print(f"An unexpected error occurred: {e}")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/text-to-speech/websocket/async_complete/main.py:93]: This is just an example.

File: deepgram/clients/speak/v1/websocket/helpers.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    Converts a URL to a WebSocket URL
+    """
+    if re.match(r"^https?://", base_url, re.IGNORECASE):
+        base_url = base_url.replace("https://", "").replace("http://", "")
+    if not re.match(r"^wss?://", base_url, re.IGNORECASE):
+        base_url = "wss://" + base_url
+    parsed_url = urlparse(base_url)
+    domain = parsed_url.netloc
+    websocket_url = urlunparse((parsed_url.scheme, domain, endpoint, "", "", ""))
+    return websocket_url
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/speak/v1/websocket/helpers.py:43]: Will revisit in the future.

File: deepgram/audio/speaker/speaker.py [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+
+        # Start the receiver thread within the start function
+        self._logger.verbose("Starting receiver thread...")
+        self._receiver_thread = threading.Thread(target=self._start_receiver)
+        self._receiver_thread.start()
+
+        self._logger.notice("start succeeded")
+        self._logger.debug("Speaker.start LEAVE")
+
+        return True
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/speaker/speaker.py:191]: Noted.

File: deepgram/audio/speaker/speaker.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+
+        self._audio = pyaudio.PyAudio()
+        self._chunk = chunk
+        self._rate = rate
+        self._format = pyaudio.paInt16
+        self._channels = channels
+        self._output_device_index = output_device_index
+
+        self._push_callback_org = push_callback
+        self._pull_callback_org = pull_callback
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/speaker/speaker.py:77]: Noted.

File: deepgram/audio/speaker/speaker.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+            self._receiver_thread.join()
+            self._receiver_thread = None  # type: ignore
+            self._logger.notice("_receiver_thread joined")
+
+        self._queue = None  # type: ignore
+
+        self._logger.notice("finish succeeded")
+        self._logger.debug("Speaker.finish LEAVE")
+
+        return True
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/audio/speaker/speaker.py:293]: Noted.

**Review by naomi-lgbt — APPROVED** [review]
Good to go!

**Pull request merged**